src/meta-srv/src/handler.rs

```diff
@@ -103,13 +114,229 @@ impl HeartbeatHandlerGroup {
+            mailbox_messages: acc.into_mailbox_messages(),
         };
         Ok(res)
     }
 }
+
+pub struct HeartbeatMailbox {
+    pushers: Arc<RwLock<BTreeMap<String, Pusher>>>,
+    sequence: Sequence,
+    senders: DashMap<MessageId, oneshot::Sender<Result<MailboxMessage>>>,
+    timeouts: DashMap<MessageId, Duration>,
+    timeout_notify: Notify,
+}
+
+impl HeartbeatMailbox {
+    pub fn create(
+        pushers: Arc<RwLock<BTreeMap<String, Pusher>>>,
+        sequence: Sequence,
+    ) -> MailboxRef {
+        let mailbox = Arc::new(Self::new(pushers, sequence));
+
+        let timeout_checker = mailbox.clone();
+        common_runtime::spawn_bg(async move {
+            timeout_checker.check_timeout_bg(1).await;
+        });
+
+        mailbox
+    }
+
+    fn new(pushers: Arc<RwLock<BTreeMap<String, Pusher>>>, sequence: Sequence) -> Self {
+        Self {
+            pushers,
+            sequence,
+            senders: DashMap::default(),
+            timeouts: DashMap::default(),
+            timeout_notify: Notify::new(),
+        }
+    }
+
+    async fn check_timeout_bg(&self, interval_millis: u64) {
+        let mut interval = tokio::time::interval(Duration::from_millis(interval_millis));
+
+        loop {
+            interval.tick().await;
+
+            if self.timeouts.is_empty() {
+                self.timeout_notify.notified().await;
+            }
+
+            let now = Duration::from_millis(common_time::util::current_time_millis() as u64);
+            let timeout_ids = self
+                .timeouts
+                .iter()
+                .filter_map(|entry| {
+                    let (id, deadline) = entry.pair();
+                    if deadline < &now {
+                        Some(*id)
+                    } else {
+                        None
+                    }
+                })
+                .collect::<Vec<_>>();
+
+            for id in timeout_ids {
+                let _ = self
+                    .on_recv(id, Err(error::MailboxTimeoutSnafu { id }.build()))
+                    .await;
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Mailbox for HeartbeatMailbox {
+    async fn send(
+        &self,
+        ch: &Channel,
+        mut msg: MailboxMessage,
+        timeout: Duration,
+    ) -> Result<MailboxReceiver> {
+        let message_id = self.sequence.next().await?;
+
+        msg.id = message_id;
+        let res = HeartbeatResponse {
+            header: None,
+            mailbox_messages: vec![msg],
+        };
+
+        let pusher_id = match ch {
+            Channel::Datanode(id) => format!("{}-{}", Role::Datanode as i32, id),
+            Channel::Frontend(id) => format!("{}-{}", Role::Frontend as i32, id),
+        };
+        let pushers = self.pushers.read().await;
+        let pusher = pushers
+            .get(&pusher_id)
+            .context(error::PusherNotFoundSnafu { pusher_id })?;
+        pusher.send(Ok(res)).await.map_err(|e| {
```

Comment:
Is it possible that the hearbeat response is received before the codes got executed below? If so, I suggest move `pusher.send` to the end of this method.

Reply:
Good catch